implementation-ios.md

```diff
@@ -0,0 +1,94 @@
+nav_titles: true
+titles_max_depth: 2
+---
+This guide is for **application developers** working to add native MPTCP support
+when running on iOS.
+
+## MPTCP connection
+
+iOS offers the possibility to establish new MPTCP connections by using its `URLSession` API, or `NSURLSession` API if you are working with Objective-C. Note that on iOS, only client side MPTCP is supported, and that the path manager can't be manually configured. However, the developers can influence the scheduler of MPTCP by selecting a type of service needed by their app.
+
+ To establish a MPTCP connection, you simply have to tweak a bit the configuration used to create the `(NS)URLSession` object. 
+
+1. Create an `(NS)URLSessionConfiguration` object, typically by using the `default` configuration. 
+2. Select the type of service you want. Available types are:
+
+    | Type of service       | Description         | 
+    |:-------------|:------------------|
+    | None     |The default service type indicating that Multipath TCP should not be used.|
+    | Handover |A Multipath TCP service that provides seamless handover between Wi-Fi and cellular in order to preserve the connection.|
+    | Interactive |A service whereby Multipath TCP attempts to use the lowest-latency interface.|
+    | Aggregate   |A service that aggregates the capacities of other Multipath options in an attempt to increase throughput and minimize latency.|
+
+    {: .note }
+    Source for available types of service: [here](https://developer.apple.com/documentation/foundation/nsurlsessionmultipathservicetype) for `NSURLSession` and [here](https://developer.apple.com/documentation/foundation/urlsessionconfiguration/2875967-multipathservicetype) for `URLSession`
+
+3. Set the `multipathServiceType` property of this configuration to the selected type of service. The value is simply `NSURLSessionMultipathServiceType${MODE_NAME}` for `NSURLSession`, and `.{MODE_NAME_LOWER}` for `URLSession`. For example, if we want to use the handover mode, we should pass `NSURLSessionMultipathServiceTypeHandover` to `NSURLSession`, and `.handover` to `URLSession`
+
+4. Create the `(NS)URLSession` by passing in the configuration to the constructor. From now on, every request done with this session should use MPTCP, if the server supports it.
+   
+{: .info }
+To enable the aggregate type of service, you'll probably also need to add the [MPTCP entitlement](https://developer.apple.com/documentation/foundation/urlsessionconfiguration/improving_network_reliability_using_multipath_tcp) to your app. For the other modes, they seem to work fine, even without this additional step, but the aggregate one crashes if you don't set it.
+
+{: .warning }
+The `(NS)URLSession` API is also available on macOS, but you won't be able to set the `multipathServiceType` in the configuration. There is thus no possible way to enable MPTCP with `(NS)URLSession` on macOS yet (though other possibilities exist, see [the FAQ](/faq.html) for more details)
```

Comment:
```suggestion
The `(NS)URLSession` API is also available on macOS, but it is no possible to set any `multipathServiceType`. There is thus not possible to enable MPTCP with `(NS)URLSession` on macOS yet.
```

With the method from the FAQ, we can create an MPTCP connection, but we cannot use multiple subflows. I didn't check myself, do not hesitate to try if you want, by using this modified version of cURL: https://github.com/matttbe/curl/tree/mptcp-macOS (`mptcp-macOS`) branch.

Reply:
I think I expressed myself badly, I wanted to say that, using the `(NS)URLSession` API, it is not possible to use MPTCP on macOS, even though it is possible to use it with other languages (C for example)

Reply:
> even though it is possible to use it with other languages (C for example)

Yes, in C it is possible, but only one path. For a multipath protocol, that's a bit annoying :)

Reply:
Indeed :D

Reply:
Oh, I was only thinking about the technique in C with `connectx()` which doesn't seem to be enough to use MPTCP. But maybe you were talking about their [Network framework](https://developer.apple.com/documentation/network)? I guess that's what is being used with [SSH](https://github.com/apple-oss-distributions/OpenSSH/blob/9b6202341ee10b42e7391229ad5c0f2eb8aea8af/openssh/sshconnect.c#L487-L489) on Apple devices (`--applemultipath`?), and it is more something for laptops, than for iOS.

If it works on macOS, maybe we should rename this page macOS, and have an introduction, saying that there are some code specifics for iOS, that works only there, then a first section for iOS, then for macOS with an example in C? WDYT?